packages/sw-background-sync-queue/src/lib/request-queue.js

```diff
@@ -0,0 +1,104 @@
+import {
+	getDb,
+	getQueue,
+	setIdbQueue,
+	createQueue,
+} from './background-sync-idb-helper';
+import {getQueueableRequest} from './queue-utils';
+import {broadcastMessage} from './broadcast-manager';
+import {
+	broadcastMessageAddedType,
+	broadcastMessageFailedType,
+	defaultQueueName,
+	tagNamePrefix,
+} from './constants';
+
+let _requestCounter = 0;
+let _queueCounter = 0;
+/**
+ * Core queue class that handles all the enqueue and dequeue
+ * as well as cleanup code for the background sync queue
+ * @class
+ */
+class Queue {
+	/**
+	 * Creates an instance of Queue.
+	 *
+	 * @memberOf Queue
+	 */
+	constructor(config, queueName) {
+		this._queueName = queueName || defaultQueueName + '_' + _queueCounter++;
+		this._config = config;
+		this._idbQHelper = getDb();
+		this._queue = getQueue(this._queueName);
+		if(!this._queue) {
+			this._queue = createQueue(this._queueName);
+		}
+	}
+
+	/**
+	 * push any request to background sync queue which would be played later
+	 * preferably when network comes back
+	 *
+	 * @param {Request} request request object to be queued by this
+	 * @param {Object} config optional config to override config params
+	 *
+	 * @memberOf Queue
+	 */
+	async push({request}) {
```

Comment:
Consider using the [`assert` library](https://github.com/GoogleChrome/sw-helpers/blob/master/lib/assert.js) to validate these parameters if they're required.